### DISCUSSION TITLE
Stride Windowing Changes

### DISCUSSION THREAD

**Discussion by Doprez** [stride3d/stride#2524] [Contributors]
[Perksey has brought up a discussion in the Discord](https://discord.com/channels/500285081265635328/1305894932682379295)  for moving things over to the Silk.NET way of Windowing and Input. I wanted to bring up all of the findings I have for our current windowing implementation here since it isn't well documented anywhere I have seen. 

With the work I was doing in trying to make the current windowing system work with any UI framework I have found a lot of possible areas of improvement we could move towards but they would be breaking changes. 
PR: https://github.com/stride3d/stride/pull/2404
Avalonia POC: https://github.com/Doprez/custom-stride-window-poc/tree/avalonia-custom-platform
Custom SDL POC: https://github.com/Doprez/custom-stride-window-poc/tree/avalonia-custom-platform/MyGame3/SDL

First thing I wanted to do was layout all of the calls and features of the current `GameWindow` class as its an easy way to point out possible issues:

## `GameWindow` Current Built-in Features
- Window title can be set through the property `Title` or the abstract method `SetTitle(string)` 
- Can be Activated or Deactivated. Seems to pause and resume window rendering? This is only used internally that I can see and is only used to stop window rendering while it is being destroyed from the `GameBase`.
This also includes public events that can be subscribed to.
- User resizing toggle can be changed through the property `AllowUserResizing` 
- The Window surface bounds are controlled by the property `ClientBounds` this is used in the `Game` class to effect the graphics device back buffer.
- The display orientation can be seen through the property `CurrentOrientation` and can also send an event when it is changed with `OnOrientationChanged`. 
- The window can be minimized with an abstract implementation and is accessed through the property `IsMinimized`.
- The window can be focused with an abstract implementation and is accessed through the property `Focused`.
- The mouse visibility can be change with the property `IsMouseVisible`.
- The window platform handle can be seen through the property `NativeWindow` as an abstract implementation.
- The window visibility can be changed through the property `Visible` as an abstract implementation.
- The window position can be changed through the property `Position`.
- The window border can be toggled through the property `IsBorderLess` as an abstract implementation.
- The window size can be changed with the property `PreferredWindowedSize`. This also seems to effect the back buffer size.
- The full screen size can be changed with the property `PreferredFullscreenSize`. This also seems to effect the back buffer size.
- Full screen borderless can be toggled with the property `FullscreenIsBorderlessWindow`.
- The window full screen can be toggled with the property `IsFullscreen`.

#### GraphicsDeviceManager.cs

- Gives information to the Window as to what orientations are supported through the internal abstract method `GameWindow.SetSupportedOrientations(DisplayOrientation)`. Does not actually effect the orientation.
- Notifies the window that the graphics device is changing the back buffer size through `Window.BeginScreenDeviceChange(bool)` and tells the window if it will be true full screen or not. Maybe used to ensure there are no locks on resources?
- Notifies the window when the graphics device is done changing the size and the resulting size through `Window.EndScreenDeviceChange(int, int)`.
- Also tells the window if the result is true full screen through the internal abstract call `Window.SetIsReallyFullscreen(bool)`.

#### GamePlatform.cs

- Handles the Run of the window by calling `GameWindow.Run()` when the `GamePlatform` is run.
- Notifies the window that it should be exiting by calling `GameWindow.Exiting`.
- Sends the windows native platform handle to the Graphics Device to render the buffer.
- Resizes the `GameWindow` when the Graphics Device has been changed using `GameWindow.Resize(int, int)`. This is likely to ensure that the backbuffer and window match or else it is handled improperly and can cause crashes if not called in order.
- Calls `GameWindow.Dispose` when the game platform has been disposed.

## Silk.NET Windowing Changes
TBD I will add more onto this but there needs to be a direct comparison to the Silk.NET windowing to address those possibilities. 

From a quick glance the Silk.NET 3.0 windowing proposal covers all that we need apart from:
- Orientation (Which isn't even used in Stride anyway)
- True full screen (Which maybe should be directly in windowing anyway)

## Stride Game Management changes
- Remove direct calls to the `GameWindow` class in the `graphicsDeviceManager` and make this be handled by the `GamePlatform` instead.
- Make the `GamePlatform` the dedicated Window Manager. This way we will need to instantiate a `GameContext` and `GameWindow` separately and then join them in the platform instead. This will help avoid the current issue of cyclical dependencies [here](https://github.com/stride3d/stride/pull/2404#issuecomment-2351341675) in the section "API improvement findings and thoughts".
- Make the window style an enum instead of multiple booleans. (Silk.NET windowing does this already if we use that). This will help make things clear as to what state the window is actually in. This is currently a problem with the way the `GraphicsDeviceManager` uses the method `Window.BeginScreenDeviceChange(bool)` since it isn't completely clear that `SetIsReallyFullscreen(bool)` is meant to be used on Graphics Device updates to tell the window it is taking over the display instead of using the window.

I'm sure there are more changes needed and I will need to go through the Silk windowing to gain a better understanding of how we could use the current one and possibly the 3.0 proposal once ready.

**Comment by Doprez** [comment]
## Silk.NET 3.0 to Replace `GameWindow`

##### replacements
- GameWindow.ClientBounds -> ISurfaceWindow.ClientArea
- GameWindow.Title -> ISurfaceWindow.Title
- GameWindow.Exiting -> ISurfaceWindow.IsCloseRequested
- GameWindow.Focused -> ISurfaceWindow.IsFocused
- GameWindow.Visible -> ISurfaceWindow.IsVisible
- GameWindow.Position -> ISurfaceWindow.Position (Int2 to Vector2)
- GameWindow.IsBorderless -> ISurfaceWindow.Border (bool to enum)
- GameWindow.AllowUserResizing -> ISurfaceWindow.Border (bool to enum)
- GameWindow.IsMinimized -> ISurfaceWindow.State (bool to enum)
- GameWindow.FullscreenIsBorderlessWindow -> ISurfaceWindow.State (bool to enum)
- GameWindow.PreferredWindowedSize -> ISurfaceWindow.Size (Int2 to Rectangle floats)

##### No direct replacements
- GameWindow.CurrentOrientation (Not used in Stride anyway AFACIT)
- GameWindow.IsMouseVisible (Should probably be in Input)
- GameWindow.PreferredFullscreenSize (Likely not meant to be part of window but maybe the Silk.NET IDisplay could improve this)
- GameWindow.NativeHandle (Could use the INativeWindow interface)

**Reply by Ethereal77** [reply]
> GameWindow.CurrentOrientation (Not used in Stride anyway AFACIT)

Maybe not used by Stride itself. But could the initial intent be that your application/game can check the orientation and react to its changes?
Like, for example, rotating elements in the UI according to the orientation the user is holding the device.

> GameWindow.IsMouseVisible (Should probably be in Input)

This may indicate if the mouse cursor is visible **when over the window** (not globally).
Right now it's not very important because there can be only one window. But maybe it would be important if we were to allow more than one window.

**Reply by Doprez** [reply]
> Maybe not used by Stride itself. But could the initial intent be that your application/game can check the orientation and react to its changes? Like, for example, rotating elements in the UI according to the orientation the user is holding the device.

I think you're right, orientation could make sense in many cases. I guess even since Windows 8 you could rotate on some devices not only mobile phones/tablets. It just needs to be implemented as it isnt right now.

> This may indicate if the mouse cursor is visible **when over the window** (not globally). Right now it's not very important because there can be only one window. But maybe it would be important if we were to allow more than one window.

Should the window instead have an event to say that the cursor is over an active window in that case and still move the actual hiding of the cursor to Stride.Input?

**Reply by Perksey** [reply]
> GameWindow.PreferredFullscreenSize (Likely not meant to be part of window but maybe the Silk.NET IDisplay could improve this)

I believe the scenarios this property would target (based on the name) is replaced by the IDisplay.VideMode/IDisplay.AvailableVideoModes API.

> GameWindow.NativeHandle (Could use the INativeWindow interface)

Yeah the intention is this is what INativeWindow is for, you'd call `TryGetPlatformInfo<Win32PlatformInfo>` for example to get the HWND.

> GameWindow.IsMouseVisible (Should probably be in Input)

This is `CursorModes` and associated APIs I think (in Silk.NET.Input, see the Multi-Backend Input proposal)

> GameWindow.CurrentOrientation (Not used in Stride anyway AFACIT)

This is a good shout.

---

I've added a `WindowedFullscreen` and renamed `Fullscreen` to `ExclusiveFullscreen`

**Reply by Ethereal77** [reply]
> Should the window instead have an event to say that the cursor is over an active window in that case and still move the actual hiding of the cursor to Stride.Input?

Something like the `MouseOver` event of WinForms? Yeah, that would do it I suppose. Although this being a visual thing that may depend on the each window, I don't know if moving out of GameWindow and into Input may obfuscate it a bit and make it difficult to find.
I mean, now Stride.Input is mainly for "reading input", so some people looking for a way to hide the cursor by first controlling if a specific window (and, if multi-window, not others) are being hovered may be confused, as that is a big departure from the current design (that comes all the way back from XNA).

**Reply by Doprez** [reply]
It's only weird to me since mouse locking is `Input.Mouse.LockPosition()` and `Input.Mouse.UnlockPosition()` but mouse hiding is `Game.IsMouseVisible` (Which accesses the window variable `GameWindow.IsMouseVisible`). It would be nice if it were consolidated into one library in my eyes. [It does look like Silks input is in the same thought process](https://github.com/dotnet/Silk.NET/blob/544b6476fc8612f9b22baa7e2cf7c31557b58dbb/documentation/proposals/Proposal%20-%20Multi-Backend%20Input.md#cursor-configuration) like Perksey said above.

**Reply by Ethereal77** [reply]
Yeah, I understand that weirdiness and I agree. Just saying that the current way is there for convenience.

Suppose we (sometime in the future) support more than one window. Then if we somehow wanted to hide the mouse in some of them but not in others, right now is just a boolean, but with the new form, the `Mouse.IsVisible` will affect all of them (or you would need to go handling `MouseOver` events and hiding / showing per window), or there would be a method `Mouse.SetVisible(Window)` or something of that kind.

In summary, what I'm saying is that it makes sense to have `Mouse.IsVisible`, but `Window.MouseVisible` is a convenience.

**Comment by Doprez** [comment]
## Game Startup issues

I have been meaning to make a quick layout of Strides current Startup process to give a visual example of the issues with making things modular. Below is a simplified graph of what I have understood, I duplicated nodes for classes to avoid making it unreadable in a graph.

![Stride-startup-spaghetti drawio](https://github.com/user-attachments/assets/e9ecdf2c-4af5-4614-88b3-94822a63b158)

Issue notes:
- Game -> GraphicsDeviceManager (Why do we have another reference for GraphicsDevice in there when we already have it in GameBase?)
- GameBase -> GameWindow (GameWindow is already part of GameContext why have both)
- GameWindow -> GameContext (If GameContext is only meant to send info to GameWindow why do we need a reference both ways)
- GameContext -> GameWindow (Same as above but maybe we leave Window creation to the GamePlatform?)
- GamPlatform -> GameBase (This feels wrong but is used to configure Services and a static method to create the relevant GamePlatform type. We should get the GamePlatform with the creation of the GameBase in the constructor instead I think.)
- GamePlatform -> GameWindow (This may be fine? Might need to be restructured but I think having the GamePlatform managing the window makes sense.)

Extra Issue notes semi unrelated:
- InputManager -> GameContext (This is only used to get the relevant input system for the platform. Its also hard coded so no one can make changes to the windowing without breaking this.)
- InputSystem -> InputManager (I dont know why this reference is needed but feels like InputManager isnt very useful to me.)


@xen2 sorry I keep tagging you in things but I think this may become semi relevant to what you were saying in the discussion about having different packages for windowing? https://github.com/stride3d/stride/discussions/2493#discussioncomment-11013897

**Reply by Ethereal77** [reply]
> Game -> GraphicsDeviceManager (Why do we have another reference for GraphicsDevice in there when we already have it in GameBase?)

`GraphicsDeviceManager` is a component inherited from XNA that allows you to have applications that aren't based in `Game`/`GameBase` but can however manage the configuration/lifetime/disposal of `GraphicsDevice`. XNA was very flexible in its design and allowed a great number of possible architectures, even if most games end up being `Game` instances.

> GameBase -> GameWindow (GameWindow is already part of GameContext why have both)

Here `GameBase` acts as a window manager. It initializes the platform[^1], creates the window, sets up the graphics device and the swap chain, and manages the lifetimes of them all.
I don't know if this is the case with Stride in particular, but I view `*Context` classes as a convenient bag of relevant information to pass around instead of a miriad of individual fields/parameters. So if this is true, `GameContext` is just the context needed for all the methods that are relevant for the functioning of the game, hence why it has "repeated" things that are already in other places.

[^1]: `GamePlatform` is an addition by Stride. It didn't exist in XNA. But I think it's an abstraction very interwined with `Game`/`GameBase`. Maybe we should thin `Game`/`GameBase` and move some of these things to `GamePlatform`.

> GameWindow -> GameContext (If GameContext is only meant to send info to GameWindow why do we need a reference both ways)

I don't know. That one is weird. `GameContext` (as in the context of a game) needs to know the game's windows. But for `GameWindow`, does it need to access the `Game`? Maybe to respond to platform events (like `Activated`, `Deactivated`, etc.)?

> GameContext -> GameWindow (Same as above but maybe we leave Window creation to the GamePlatform?)

I think it makes more sense if the window creation is the responsibility of the platform, yes.

> GamePlatform -> GameBase (This feels wrong but is used to configure Services and a static method to create the relevant GamePlatform type. We should get the GamePlatform with the creation of the GameBase in the constructor instead I think.)

The problem here arises because `GamePlatform` should be an abstraction for the platform (i.e. the system, HW & SW). But right now, it's the game instance the one that "creates" the platform and configures it, so there's a very strange back and forth between the two.

> GamePlatform -> GameWindow (This may be fine? Might need to be restructured but I think having the GamePlatform managing the window makes sense.)

Yeah, `GamePlatform` as a window manager makes more sense.

> InputSystem -> InputManager (I dont know why this reference is needed but feels like InputManager isnt very useful to me.)

~Another remnant of XNA I think~. `InputManager` allows you to have input devices without the `Game` infrastructure. `InputSystem` is the Stride's way to encapsulate it. In other words, `InputSystem` is a system of the engine living in the engine's assembly, while `InputManager` is the component specialized in input handling (the platform-specific implementation I suppose) living in `Stride.Input` assembly.
 
Anyway, I hope this is all clear, I'm not talking nonsense and I haven't confused you more. The history of many of these systems is pretty old and convoluted at this point. 😁

PS: If you want to go down the rabbit hold, you can check the way XNA was structured. You'll find more than one familiar thing there. I haven't been able to find the original MSDN docs, but the docs for [MonoGame](https://docs.monogame.net/api/Microsoft.Xna.Framework.Game.html) are sufficient, as it was an almost one-to-one reimplementation of XNA.

**Reply by Doprez** [reply]
This does help a lot, thank you! I definitely did notice similarities especially in naming between Stride and MonoGame so Ill take a look at their docs before I try to look into a plan in case I start reinventing the wheel. 

`GameContext` is especially weird right now because it seems to be related more to a window context as that's all its currently used for. I do agree with what you are saying though, having a context to give initialization parameters to the `GameBase` makes sense in my eyes.

 I think I also agree with the idea of moving more responsibility to the `GamePlatform` especially for any `GameSystems` that may not be fully cross platform. I don't know if there are any specific examples currently in `Game`/`GameBase` but I could already think of having a specific `GamePlatform` for VR and removing the hard coded references from `Game` that are created on initialization, should help with trimming unneeded libraries.

I guess the other thing I would be curious of is how these changes may be able to help with more dynamic modules? Ill go through MonoGames docs first to see if they have any guidance but I could see `GameContext` being useful there.

**Reply by Ethereal77** [reply]
> I think I also agree with the idea of moving more responsibility to the GamePlatform especially for any GameSystems that may not be fully cross platform.

The original design of XNA divided the concept of `Game`, which represents the application itself, from the `GameComponents`, that were components or systems that run alongside the application, register services that can be queried through a service registry (like a primitive DI container), receive events, are updated each tick (`IUpdateable`) or drawn each frame (`IDrawable`).

As the purpose of `GameComponents` is to be more or less self-contained and independent of the application itself (at least in theory), maybe we could ask the users of code-only Stride. After all, XNA was a code-only framework first.

If we could cleanly separate the code-only framework parts (like those of XNA heritage) from the engine parts (the added Stride infrastructure and systems), both parts would be cleaner and simpler to reason about.

> I don't know if there are any specific examples currently in Game/GameBase but I could already think of having a specific GamePlatform for VR and removing the hard coded references from Game that are created on initialization, should help with trimming unneeded libraries.

That would be nice. Notice however that VR can be used in "any" platform and with different graphics APIs (DX, Vulkan, etc). So a VR platform is not really a platform.

**Reply by ericwj** [reply]
I think one useful mindset to view all this with is software engineering principles such as the Single Responsibility Principle. XNA must've been designed by bright minds in this area and with a mission much, much broader than that Stride probably now has, if you would describe it in one sentence. Therefore, perhaps over the years Stride might have become to violate some of these principles, if only because people mostly saw it in a very restricted way: as a game engine. It could be much more, for many more people. Hence besides looking at XNA, I would recommend refreshing your active recollection of the 10 or so most important principles and rigorously shaving cutting refactoring to get Stride back to a clean, lean, mean software engineering design applicable to the broadest audience due to the flexibility those principles provide alone. Becoming code only, able to be used with DI, for the construction phase at least, is a very good sign of being on the right track.

**Reply by ericwj** [reply]
KISS, DRY, SOLID, Occams Razor, PLA are the most important. Single Responsibility Principle is part of SOLID. Here's the first link I found: [Principles of Software Engineering](https://www.interviewbit.com/blog/principles-of-software-engineering/)

**Reply by Doprez** [reply]
I agree, I think the main thing I am struggling to work out is specifically the single responsibility while working with the current structure. 

From what I understood for the GamePlatform is that it is meant to determine the following things:
- Supported graphics backends (Determined by the project build settings so not used here)
- Graphics device
- The host window 
- OS specific features

The thing I am struggling to figure out is how that is any better than having that same functionality managed by the `GameBase` class.

The other thing I thought of was maybe the purpose of the GamePlatform should purely be used to find the best features for the platform its running on? That way the class becomes much more focused and leaves the rest of the running and management features available through the GameBase class. It also becomes an optional utility rather than a seemingly extra layer with no major benefit.

I think the rest can make sense with those principles in mind: 
- `GameContext` can contain Feature specific information including the currently configured platform settings
- `GameSettings` stays as is but I could add a list of `GameSystem`s that can be added and removed to allow for the Game Studio to include systems dynamically rather than having it hard coded in the overridden `Game` class
- `GameBase` stays as is depending on what to do with the `GamePlatform`. If `GamePlatform` changes we may need to move window running to the `GameBase` Run method rather than leaving it in the `GamePlatform` Run method which seems unnecessary already. 
- `GameSystem`s stays as is, I may need to add a data contract to the class to allow Game Studio to add it to the `GameSettings` but it already takes in the `IServiceProvider` in its constructor(will need to see what can be done with the serializer since it needs an empty contructor).
- `GameWindow` stays as is, ignoring the possible change with Silk.Windowing I think this class only changes in the sense of what should reference and manage it. Either the GamePlatform or the GameBase.
- `InputManager` will have the `GameContext` reference removed and will instead be configured through `Game`. This solves a separate issue where the input had to be configured through Stride source rather then being allowed to add your own at runtime.

I think that covers most of the issues? At least with removing seemingly random references being everywhere and having a central point of truth. The last thing I need to look into is the cyclic reference between `GameWindow` and `GameContext` but I think with the above things getting figured out that will become easier to what should be managing the window.

**Reply by ericwj** [reply]
I can't comment very specifically, since I haven't had the code cloned for months and I was never intimately known with this part of Stride. But I can say a few things.

* It could well be that `GamePlatform` fulfills the role of a very specific, limited integrator.  
That is why it might be combining these things and have references to other things which themselves have a single responsibility.  
Once you start doing something useful or hard, like rendering, you might need to integrate multiple things. There is a tradeoff between utility and SRP in such specific circumstances.

* How `GamePlatform` relates to `GameBase`, without looking in the code, is probably that `GamePlatform` isn't "app code" and `GameBase` is meant to be the primary extension point for writing "app code". An "app", meaning what you can create deriving from `GameBase`, is just the extreme form of "integrator". A big pile of stuff, often mutually unrelated stuff, composed together to provide specific features and behavior. Usually not at all reusable.  

* Notice I didn't say "executable", since it isn't required to be - Stride can be a component still in a larger app. Ideally possibly more than one such Stride. And I didn't say "window", or "surface" because even that is optional. Stride could make video's, or just images, as a server without UI, if it was really code only. Lets say you crank up detail, lighting, insanely, and do ray tracing, for the in-game video, which you ship as .mp4. Or replays; no UI, but 10 camera's, keyboards and mice, one for each player.  
So you don't have to premeditate that, but people can dream this sort of thing up if only Stride is flexible enough, through sound application of software engineering principles.

* So yes, windowing should be a Stride thing, not a Silk thing. Definitely.

* Also useful is to look at some very simple examples lets say in the [DirectX Graphics Samples](https://github.com/microsoft/DirectX-Graphics-Samples) repository. Even if Stride is much more than a simple sample, the underlying basic structure of a rendering application should be there and possibly be available for replacement. Most samples have the basic structure:
    1. Create device dependent resources
    1. Create size dependent resources
    1. Listen for device lost to redo step 1
    1. Listen for resize to redo step 2
    1. Update/Render loop

    Which means the primitives absolutely required to start Stride should not be much more than that. I don't mean what you get when you make a game in the editor, but I mean what you need when you would create those objects relevant here yourself in say a console app. Maybe you can scrutinize what you have found in that light?

* This is one of the worst things ever happening to Stride imho. The `#if` and configuring the backend through MSBuild.  
    > * Supported graphics backends (Determined by the project build settings so not used here)  

    The wanted state is that you can ship multiple backends and construct the desired one at runtime (through DI). Would you see opportunity to make this part of Stride more suitable to that end, I think that would be awesome.

**Reply by Ethereal77** [reply]
@ericwj I'm not saying XNA was design by the brightest minds and we should follow it to the letter. I'm saying XNA inspired SharpDX Toolkit, which then inspired Stride (Paradox back then), and that is a pretty flexible architecture for code-only (as that was the intent of XNA).

Also, as we currently have code-only users in the community and people who don't even like to launch Game Studio, let's not forget that. People even have started to use Stride for this because of lack of flexibility in other engines.

Right now you can say Stride is composed of several very different parts:
1. A game framework, the part that was inspired by XNA, etc. This includes things like `Stride.Core`, `Stride.Core.Mathematics`, or `Stride.Graphics`.
2. A game engine, the part that using the game framework compounds a game with its rendering pipeline, interaction, audio, streaming, etc.
3. A game content pipeline for consuming and compiling content, including also an editor (Game Studio).

I'm all in for improving and simplifying Stride where needed, but please don't destroy the core principles that make it so powerful and flexible at the same time. Just my opinion, of course.

**Reply by Doprez** [reply]
As someone who primarily uses the code only library and only opens Game Studio for scene editing. That was the main reasons I even decided to look into this in the first place. 

The issues I have dealt with for using code only are:
- locked down APIs that didnt allow for custom window implementations.
- internal APIs that seem incomplete (probably why they're internal).
- internal calls in the Game class that prevented [manual fixes to bugs](https://github.com/stride3d/stride/pull/2274#discussion_r1615324606).
- The content pipeline is painful for code only without a ton of added code https://github.com/stride3d/stride-community-toolkit/issues/176

Definitely not my goal to change any of the core functions that make Stride flexible. Ideally nothing should change to the average users of Stride in terms of functionality or current access. The majority of this work is more about making things even more flexible by decoupling things that aren't needed by everyone like the `GameWindow` for server run games or Systems that are not configurable without rewriting the entire `Game` class which you can't even do because of internal classes.

The changes I want to make are for people who want:
- more customizable game startups
- to make their own [UI framework implementations](https://github.com/stride3d/stride/discussions/2505)
- options to add systems easily at start

Most of the changes I have in mind are all within the Stride.Games library and should touch nothing else except to update any GameSystems, EntityProcessors, EntityComponents, etc to use the new references and even most of those are using the `IServiceRegistry` so a lot should already be OK.

**Reply by ericwj** [reply]
@Ethereal77 sure, I am first of all trying to make clear what the core principles even are, or how they came about. What I said about refactoring was towards cleanup of core principles where they might have become violated and improving flexibility, not at all for destruction.

I haven't commented on concrete things in Stride, or compared with XNA, because I don't have the knowledge - that's more you, reading your comments. Very valuable. We come at it from entirely different angles with the goal of improving understanding. Mine being almost entirely abstract.

Other than expecting such a Microsoft project to be meticulously designed and reviewed by at least very decent architects, I have no opinion on the quality of XNA, or its problems. But I like what you say - that one of its core design principles was code-first. However that works though, the other parts of Stride will have to remain able to deal with it. I don't think that is much of a restriction though.
